org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java

```diff
@@ -0,0 +1,158 @@
+package org.eclipse.buildship.core.workspace.internal;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Set;
+
+import org.gradle.tooling.BuildLauncher;
+import org.gradle.tooling.CancellationToken;
+import org.gradle.tooling.GradleConnectionException;
+import org.gradle.tooling.ProgressListener;
+import org.gradle.tooling.ProjectConnection;
+import org.gradle.tooling.ResultHandler;
+import org.gradle.tooling.events.OperationType;
+import org.gradle.tooling.model.Launchable;
+import org.gradle.tooling.model.Task;
+
+import com.gradleware.tooling.toolingmodel.repository.FixedRequestAttributes;
+import com.gradleware.tooling.toolingmodel.repository.TransientRequestAttributes;
+
+final class ConnectionAwareBuildLauncherDelegate extends ConnectionAwareLongRunningOperationDelegate<BuildLauncher> implements BuildLauncher {
```

Comment:
I think you could save a lot of code here by using `java.lang.Proxy` instead of implementing the delegation pattern by hand. The reflection overhead would be negligible, since we don't call these methods in a hot loop.